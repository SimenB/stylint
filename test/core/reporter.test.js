'use strict';

const _ = require('lodash');
const reporter = require('../../src/core/reporter');

const countSeverities = require('../../src/utils/countSeveritiesInMessages');

const stylint = require('../../index');

function genMessage(filePath, ruleIds, severity) {
  const ruleIdsArray = Array.isArray(ruleIds) ? ruleIds : [ruleIds];

  return {
    filePath,
    messages: ruleIdsArray.map(ruleId => ({
      line: 1,
      column: -1,
      severity,
      message: 'This is not OK',
      source: '',
      ruleId,
    })),
  };
}

function genWarning(filePath, rule) {
  return genMessage(filePath, rule, 'warning');
}

function genError(filePath, rule) {
  return genMessage(filePath, rule, 'error');
}

function generateReport(result) {
  const severities = countSeverities(_.flatMap(result, 'messages'));

  severities.results = result || [];

  return severities;
}

describe('reporter', () => {
  it('should have correct output on no message', () => {
    expect(reporter(generateReport())).toMatchSnapshot();
  });

  it('should include kill message', () => {
    expect(reporter(generateReport([genWarning('some file.styl', 'no-undefined')]), {}, true)).toMatchSnapshot();
  });

  it('should include max errors and max warnings', () => {
    expect(
      reporter(generateReport([genWarning('some file.styl', 'no-undefined')]), { maxErrors: 5, maxWarnings: 5 })
    ).toMatchSnapshot();
  });

  it('should skip non-valid max errors and max warnings', () => {
    expect(
      reporter(generateReport([genWarning('some file.styl', 'no-undefined')]), { maxErrors: -1, maxWarnings: 5 })
    ).toMatchSnapshot();
    expect(
      reporter(generateReport([genWarning('some file.styl', 'no-undefined')]), { maxWarnings: 5 })
    ).toMatchSnapshot();
    expect(
      reporter(generateReport([genWarning('some file.styl', 'no-undefined')]), { maxErrors: 2 })
    ).toMatchSnapshot();
  });

  it('should format warning correctly', () => {
    expect(reporter(generateReport([genWarning('some file.styl', 'no-undefined')]))).toMatchSnapshot();
  });

  it('should format error correctly', () => {
    expect(reporter(generateReport([genError('some file.styl', 'no-undefined')]))).toMatchSnapshot();
  });

  it('should format error and warning correctly', () => {
    const error = genError('some file.styl', 'no-undefined');
    const warning = genWarning('some file.styl', 'no-undefined');

    expect(reporter(generateReport([error, warning]))).toMatchSnapshot();
  });

  it('should format column', () => {
    const error = genError('some file.styl', 'no-undefined');

    error.messages[0].column = 5;

    expect(reporter(generateReport([error]))).toMatchSnapshot();
  });

  it('should not group files by default', () => {
    const error1 = genError('some file.styl', 'no-undefined');
    const error2 = genError('some file.styl', 'no-undefined');
    const error3 = genError('some other file.styl', 'no-undefined');

    expect(reporter(generateReport([error1, error2, error3]))).toMatchSnapshot();
  });

  it('should group files correctly', () => {
    const error1 = genError('some file.styl', ['no-undefined', 'no-undefined']);
    const error2 = genError('some other file.styl', 'no-undefined');

    expect(reporter(generateReport([error1, error2]), { groupOutputByFile: true })).toMatchSnapshot();
  });
});

describe('(Old tests) Reporter should: ', () => {
  const app = stylint().create();

  it('return correctly formatted msg', () => {
    app.cache.rule = 'universal';

    const msg = {
      filePath: 'testReporter',
      messages: [
        {
          message: 'universal disallowed',
          severity: 'warning',
          line: 1,
          column: 0,
          ruleId: 'universal',
          source: 'Reporter Lyfe*',
        },
      ],
    };

    expect(app.reporter({ results: [msg], errorCount: 0, warningCount: 1 })).toMatchSnapshot();
  });
});
