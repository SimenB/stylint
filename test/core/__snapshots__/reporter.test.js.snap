// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Old tests) Reporter should:  return correctly formatted msg 1`] = `
"[4mtestReporter[24m
1 [90muniversal[39m [33mwarning[39m universal disallowed

Stylint: 0 Errors.
Stylint: 1 Warnings."
`;

exports[`formatter should format column 1`] = `
"[4msome file.styl[24m
1:5 [90mno-undefined[39m [31merror[39m This is not OK

Stylint: 1 Errors.
Stylint: 0 Warnings."
`;

exports[`formatter should format error and warning correctly 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [31merror[39m This is not OK

[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 1 Errors.
Stylint: 1 Warnings."
`;

exports[`formatter should format error correctly 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [31merror[39m This is not OK

Stylint: 1 Errors.
Stylint: 0 Warnings."
`;

exports[`formatter should format warning correctly 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 0 Errors.
Stylint: 1 Warnings."
`;

exports[`formatter should group files correctly 1`] = `
"some file.styl
FILE           LINEDATA SEVERITY MESSAGE        RULE        
[4msome file.styl[24m 1        [31merror[39m    This is not OK [90mno-undefined[39m
[4msome file.styl[24m 1        [31merror[39m    This is not OK [90mno-undefined[39m

some other file.styl
FILE                 LINEDATA SEVERITY MESSAGE        RULE        
[4msome other file.styl[24m 1        [31merror[39m    This is not OK [90mno-undefined[39m

Stylint: 3 Errors.
Stylint: 0 Warnings."
`;

exports[`formatter should have correct output on no message 1`] = `""`;

exports[`formatter should include kill message 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 0 Errors.
Stylint: 1 Warnings.
Stylint: Over Error or Warning Limit."
`;

exports[`formatter should include max errors and max warnings 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 0 Errors. (Max Errors: 5)
Stylint: 1 Warnings. (Max Warnings: 5)"
`;

exports[`formatter should not group files by default 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [31merror[39m This is not OK

[4msome file.styl[24m
1 [90mno-undefined[39m [31merror[39m This is not OK

[4msome other file.styl[24m
1 [90mno-undefined[39m [31merror[39m This is not OK

Stylint: 3 Errors.
Stylint: 0 Warnings."
`;

exports[`formatter should skip non-valid max errors and max warnings 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 0 Errors.
Stylint: 1 Warnings. (Max Warnings: 5)"
`;

exports[`formatter should skip non-valid max errors and max warnings 2`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 0 Errors.
Stylint: 1 Warnings. (Max Warnings: 5)"
`;

exports[`formatter should skip non-valid max errors and max warnings 3`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 0 Errors. (Max Errors: 2)
Stylint: 1 Warnings."
`;

exports[`reporter should format column 1`] = `
"[4msome file.styl[24m
1:5 [90mno-undefined[39m [31merror[39m This is not OK

Stylint: 1 Errors.
Stylint: 0 Warnings."
`;

exports[`reporter should format error and warning correctly 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [31merror[39m This is not OK

[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 1 Errors.
Stylint: 1 Warnings."
`;

exports[`reporter should format error correctly 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [31merror[39m This is not OK

Stylint: 1 Errors.
Stylint: 0 Warnings."
`;

exports[`reporter should format warning correctly 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 0 Errors.
Stylint: 1 Warnings."
`;

exports[`reporter should group files correctly 1`] = `
"some file.styl
FILE           LINEDATA SEVERITY MESSAGE        RULE        
[4msome file.styl[24m 1        [31merror[39m    This is not OK [90mno-undefined[39m
[4msome file.styl[24m 1        [31merror[39m    This is not OK [90mno-undefined[39m

some other file.styl
FILE                 LINEDATA SEVERITY MESSAGE        RULE        
[4msome other file.styl[24m 1        [31merror[39m    This is not OK [90mno-undefined[39m

Stylint: 3 Errors.
Stylint: 0 Warnings."
`;

exports[`reporter should have correct output on no message 1`] = `""`;

exports[`reporter should include kill message 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 0 Errors.
Stylint: 1 Warnings.
Stylint: Over Error or Warning Limit."
`;

exports[`reporter should include max errors and max warnings 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 0 Errors. (Max Errors: 5)
Stylint: 1 Warnings. (Max Warnings: 5)"
`;

exports[`reporter should not group files by default 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [31merror[39m This is not OK

[4msome file.styl[24m
1 [90mno-undefined[39m [31merror[39m This is not OK

[4msome other file.styl[24m
1 [90mno-undefined[39m [31merror[39m This is not OK

Stylint: 3 Errors.
Stylint: 0 Warnings."
`;

exports[`reporter should skip non-valid max errors and max warnings 1`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 0 Errors.
Stylint: 1 Warnings. (Max Warnings: 5)"
`;

exports[`reporter should skip non-valid max errors and max warnings 2`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 0 Errors.
Stylint: 1 Warnings. (Max Warnings: 5)"
`;

exports[`reporter should skip non-valid max errors and max warnings 3`] = `
"[4msome file.styl[24m
1 [90mno-undefined[39m [33mwarning[39m This is not OK

Stylint: 0 Errors. (Max Errors: 2)
Stylint: 1 Warnings."
`;
